tests/test_bonded.py

```diff
@@ -12,54 +12,110 @@
 
 class TestBonded(GradientTest):
 
-    # def test_centroid_restraint(self, n_particles=10, n_A=4, n_B=3, kb=5.4, b0=2.3):
-    #     """Randomly define subsets A and B of a larger collection of particles,
-    #     generate a centroid restraint between A and B, and then validate the resulting CentroidRestraint force"""
-    #     box = np.eye(3) * 100
-
-    #     # specific to centroid restraint force
-    #     relative_tolerance_at_precision = {np.float32: 2e-5, np.float64: 1e-9}
-
-    #     for precision, rtol in relative_tolerance_at_precision.items():
-    #         x_primal = self.get_random_coords(n_particles, 3)
-
-    #         gai = np.random.randint(0, n_particles, n_A, dtype=np.int32)
-    #         gbi = np.random.randint(0, n_particles, n_B, dtype=np.int32)
-
-    #         masses = np.random.rand(n_particles)
-
-    #         ref_nrg = jax.partial(
-    #             bonded.centroid_restraint,
-    #             masses=masses,
-    #             group_a_idxs=gai,
-    #             group_b_idxs=gbi,
-    #             kb=kb,
-    #             b0=b0
-    #         )
-
-    #         # we need to clear the du_dp buffer each time, so we need
-    #         # to instantiate test_nrg inside here
-    #         test_nrg = potentials.CentroidRestraint(
-    #             gai,
-    #             gbi,
-    #             masses,
-    #             kb,
-    #             b0
-    #         )
-
-    #         params = np.array([], dtype=np.float64)
-    #         lamb = 0.3  # doesn't matter
-
-    #         self.compare_forces(
-    #             x_primal,
-    #             params,
-    #             box,
-    #             lamb,
-    #             ref_nrg,
-    #             test_nrg,
-    #             rtol,
-    #             precision=precision
-    #         )
+    def test_centroid_restraint(self, n_particles=10, n_A=4, n_B=3, kb=5.4, b0=2.3):
+        """Randomly define subsets A and B of a larger collection of particles,
+        generate a centroid restraint between A and B, and then validate the resulting CentroidRestraint force"""
+        box = np.eye(3) * 100
+
+        # specific to centroid restraint force
+        relative_tolerance_at_precision = {np.float32: 2e-5, np.float64: 1e-9}
+
+        for precision, rtol in relative_tolerance_at_precision.items():
+            x_primal = self.get_random_coords(n_particles, 3)
+
+            gai = np.random.randint(0, n_particles, n_A, dtype=np.int32)
```

Comment:
Note this will sample with replacement `n_A` times, so it's possible for `gai` to contain fewer than `n_A` unique points

Reply:
Yes - this is intentional, since duplicate indices may appear in the RMSD Restraint.